server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -206,18 +206,58 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
          * We *could* detect these cases and only fork in then, but that is complex
          * to get right and the fork is fairly low overhead.
          */
+        final ClusterState initialState = clusterService.state();
         final int indexingOps = bulkRequest.numberOfActions();
         final long indexingBytes = bulkRequest.ramBytesUsed();
-        final boolean isOnlySystem = isOnlySystem(bulkRequest, clusterService.state().metadata().getIndicesLookup(), systemIndices);
-        final Releasable releasable = indexingPressure.markCoordinatingOperationStarted(indexingOps, indexingBytes, isOnlySystem);
-        final ActionListener<BulkResponse> releasingListener = ActionListener.runBefore(listener, releasable::close);
+        final boolean isOnlySystem = isOnlySystem(bulkRequest, initialState.metadata().getIndicesLookup(), systemIndices);
         final String executorName = isOnlySystem ? Names.SYSTEM_WRITE : Names.WRITE;
-        threadPool.executor(Names.WRITE).execute(new ActionRunnable<>(releasingListener) {
-            @Override
-            protected void doRun() {
-                doInternalExecute(task, bulkRequest, executorName, releasingListener);
+        final Releasable releasable = indexingPressure.markCoordinatingOperationStarted(indexingOps, indexingBytes, isOnlySystem);
+        // We should use the releasingListener from here onwards
+        final ActionListener<BulkResponse> releasingListener = ActionListener.runBefore(originalListener, releasable::close);
```

Comment:
I have a slight preference for leaving these lines in the same order to simplify the diff, unless I'm missing something about the new order being required?

Reply:
It was accidental. Should be reverted by now.